### PR TITLE
Feature/869 stream updates

### LINKED ISSUE #1545: StreamReader::readString() should read characters, not bytes
invisible = unreleased code

all string operations in Qore read characters where possible, when the encoding is known - this should be the same with StreamReader

### LINKED ISSUE #1546: implement variable buffer size support for BufferedStreamReader

### LINKED ISSUE #1547: File::getchar() does not read in a character with multi-byte character encodings

### LINKED ISSUE #1548: File::read() should respect character semantics instead of byte semantics
with byte semantics this method could return an invalid string with multi-byte character encodings

### PR DESCRIPTION
**Pull request by davidnich**
closes #1545 
closes #1546 
closes #1547 
closes #1548

### CODE REVIEW

File: include/qore/intern/BufferedStreamReader.h [hunk 1]
```diff
@@ -90,24 +109,69 @@ class BufferedStreamReader : public StreamReader {
...
-            memmove(buf, p + eolLen, bufCount);
+            if (bufCount)
+               memmove(buf, p + eolLen, bufCount);
             return line.release();
          }
          else {
             if (rc > 0) {
-               qore_size_t dataSize = (bufCount >= eolLen) ? bufCount - eolLen : 0;
+               if (eol) {
+                  // we read in new data, so here we need to look for a partial eol match at the end of the buffer
```
Comment (omusil24) [inline include/qore/intern/BufferedStreamReader.h]: @davidnich Why did you add so much new code here? The previous code worked just fine, including correctly handling EOL divided by the edge of the buffer.

Comment (davidnich) [reply include/qore/intern/BufferedStreamReader.h]: because it didn't work - if you don't believe me, then run the new tests on the old implementation and see what happens - the main case that was not covered is when an `eol` string was passed and that string spanned a buffer boundary.  also there were quite a few fixes regarding the variable buffer size that were only visible when you select a large buffer size - they were hard to test with the hardcoded 4K buffer size

Comment (omusil24) [reply include/qore/intern/BufferedStreamReader.h]: The code works fine including the new tests unless you select stupidly small buffer size like 1 or 2 which defeats the purpose of buffering (performance) anyway (the algorithm was not designed for that). Talking about the BufferedStreamReader here, just to be clear.

Comment (davidnich) [reply include/qore/intern/BufferedStreamReader.h]: The old tests for the case when an explicit multi-character eol string spanned a buffer were incorrect - the eol string did not span the buffer.  I have just pushed new tests now that test this case correctly, and the old code is now failing in stream-reader.qtest, while the new code is passing in buffered-stream-reader.qtest.

The small buffer size is not "stupid"; you might say it's unrealistic, but first of all it's not our job to determine what arguments Qore users will use to functions, and in any case Qore needs to work properly and predictably with all potential arguments to all functions/methods etc.  Also the small buffer sizes made it very easy to test the edge cases.

Our BufferedStreamReader class is modeled after Java's  [BufferedReader](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html) which also has a constructor that takes a buffer size, so the buffer size argument is sensible in my opinion.

Please fix the bug in the unbuffered reader as well before merging.

**Comment by davidnich** [comment]
@omusil24 I think we should take some time and fix these issues correctly while also improving the internal stream implementation for *StreamReader - in my opinion, BufferedStreamReader should only implement a buffered read function and not have to override the other methods - in other words the only difference between the internal implementations of the `StreamReader` and `BufferedStreamReader` C++ classes should be the buffer handling.

Also I believe we should make a helper class for the string reading functionality now duplicated in `qore_qf_private::readString()` and `BufferedStringReader::readString()` to avoid code duplication.

Let discuss on Monday.

**Comment by davidnich** [comment]
@omusil24 now it should be ready for merging - never mind - still needs a little work...
